pdns/dnsdist-lua-bindings-dnsquestion.cc

```diff
@@ -54,6 +54,27 @@ void setupLuaBindingsDNSQuestion()
   g_lua.registerFunction<bool(DNSQuestion::*)()>("getDO", [](const DNSQuestion& dq) {
       return getEDNSZ((const char*)dq.dh, dq.len) & EDNS_HEADER_FLAG_DO;
     });
+  g_lua.registerFunction<vector<uint8_t>(DNSQuestion::*)(void)>("getTrailingData", [](const DNSQuestion& dq) {
+      const uint8_t* message = reinterpret_cast<const uint8_t*>(dq.dh);
+      const uint16_t length = getDNSPacketLength(reinterpret_cast<const char*>(message), dq.len);
+      vector<uint8_t> tail(message + length, message + dq.len);
+      return tail;
+    });
+  g_lua.registerFunction<bool(DNSQuestion::*)(vector<pair<int, uint8_t>>)>("setTrailingData", [](DNSQuestion& dq, const vector<pair<int, uint8_t>>&data) {
+      uint8_t* message = reinterpret_cast<uint8_t*>(dq.dh);
+      const uint16_t length = getDNSPacketLength(reinterpret_cast<const char*>(message), dq.len);
+      if(length + data.size() > dq.size) {
+        return false;
+      }
+
+      /* Copy data from the Lua array, whose first index is 1 instead of 0. */
+      dq.len = length + data.size();
+      uint8_t* tail = message + length - 1;
+      for(const auto& pair : data) {
+        *(tail + pair.first) = pair.second;
```

Comment:
I'm probably paranoid but I'd prefer the use of a counter instead of using the indexes passed by `Lua`.

Reply:
I was also wondering about that... but is the order of data from Lua tables guaranteed? And if not, should I just guard the write like e.g.
```cpp
int size = data.size();
for(const auto& pair : data) {
  if(pair.first < 1 || pair.first > size) {
    /* reject sparse input */
    return false;
  }
  *(tail + pair.first) = pair.second;
}
```

Reply:
To be fair if the order is not right I'd consider it a bug in the wrapper. Still I'd like some sort of check, yes, but we need to be careful because we already increased `dq.len` at that point.

Reply:
So what kind of protection would you like to see?

Reply:
@rgacogne would it make more sense to follow what #6068 is doing with EDNS option data and expose the trailing data as a string rather than a uint8 list? It would work as long as 0x00 is allowed and strings are sufficiently manipulable in Lua, and would resolve the ordering concern.

Reply:
Sorry it took me so long to answer this comment :-/
Yes, we have been using binary-safe strings in dnsdist for this kind of things so far, so IMHO it would be a better choice. Again, thanks a lot for contributing.